internal/internal_workflow.go

```diff
@@ -1742,3 +1754,56 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
 func (us updateSchedulerImpl) Yield(ctx Context, reason string) {
 	getState(ctx).yield(reason)
 }
+
+func (m *mutexImpl) Lock(ctx Context) error {
+	err := Await(ctx, func() bool {
+		return !m.locked
+	})
+	if err != nil {
+		return err
+	}
+	m.locked = true
+	return nil
+}
+
+func (m *mutexImpl) TryLock(ctx Context) bool {
```

Comment:
Arguably don't need a context here, but I don't mind it, so long as it's understood that calling `Lock` with a cancelled context is different than calling `TryLock` with a cancelled context.

Reply:
Ah I added a context here so I could add a check this wasn't being called in an read only context.